lib/primitives/mtx.js

```diff
@@ -1881,6 +1890,132 @@ class CoinSelector {
     this.fee = Math.min(this.hardFee, CoinSelector.MAX_FEE);
     this.fund();
   }
+
+  /**
+   * Initiate selection using new selection algorithm.
+   */
+
+  async selectByValue() {
+    // add dummy output for change.
+    const change = new Output();
+
+    if (this.changeAddress) {
+      change.script.fromAddress(this.changeAddress);
+    } else {
+      // In case we don't have a change address,
+      // we use a fake p2pkh output to gauge size.
+      change.script.fromPubkeyhash(Buffer.allocUnsafe(20));
+    }
+
+    this.tx.outputs.push(change);
+
+    let target = this.outputValue;
+
+    // calculate fee for everything except inputs.
+    let size = await this.tx.estimateSize();
+    const fee = this.getFee(size);
+    // add fee to target value.
+    target += fee;
+
+    // begin selection.
+
+    // while target is greater than
+    // the largest coin we have, we
+    // will keep selecting the largest coin
+
+    this.index = this.coins.length - 1;
+    let coin = this.coins[this.index];
+    let value = Amount.value(coin.value);
+    let spendingSize = await coin.estimateSpendingSize(this.getAccount);
+    let effectiveValue = value - this.getFee(spendingSize);
+    while (target >= effectiveValue) {
+      this.tx.addCoin(coin);
+      this.chosen.push(coin);
+
+      target -= effectiveValue;
+
+      this.index--;
+      if (this.index < 0)
+        return;
+
+      coin = this.coins[this.index];
+      value = Amount.value(coin.value);
+      spendingSize = await coin.estimateSpendingSize(this.getAccount);
+      effectiveValue = value - this.getFee(spendingSize);
```

Comment:
I think this code is repeated a few times -- can we have a `getEffectiveValue()` function somewhere? Maybe it lives in `Coin` class?

Reply:
If an instance of `Coin` knows what their input weight is, a `getEffectiveValue(feerate)` function would do.